tools/flang2/flang2exe/dinit.cpp

```diff
@@ -580,7 +580,8 @@ dinit_varref(VAR *ivl, SPTR member, CONST *ict, DTYPE dtype,
       *repeat = 0;
       (void)dinit_varref(ivl, member, ict->subc, dtype, struct_bytes_initd,
                          repeat, base_off);
-      i = *repeat = ad_val_of(AD_NUMELM(AD_DPTR(ict->dtype)));
```

Comment:
The previous definition of num_elem looks similar. What is the difference in value?
```
num_elem = ad_val_of(AD_NUMELM(AD_PTR(init_sym)));
```

Reply:
`num_elem` is 10 in the earlier call and 4 (or whatever the value of `n` in the reproducer is) in the one I removed here (same as for `flang1`).

Reply:
Which of the three paths above would flang take for an array constructor with an implied do loop? If it is the last block, where `num_elem = ad_val_of(AD_NUMELM(AD_PTR(init_sym)))`, what is `init_sym` set to? Also, it would be nice to add a comment in that block to indicate it is used for handling implied do loops.

Reply:
Having another look at this code, I am starting to think I might have not understood how it is meant to work correctly...

But answering your questions, @bryanpkc ...
flang2 enters the [first block](https://github.com/flang-compiler/flang/blob/master/tools/flang2/flang2exe/dinit.cpp#L484-L509) . We have `sptr` set to the `array` variable. 
`numelem` is set to 10 and `elemsize` to 4.

Then we [recursively run the same](https://github.com/flang-compiler/flang/blob/8afbdfb4b16e77e40b332b0fc803ebbe4e266c4a/tools/flang2/flang2exe/dinit.cpp#L581) function for the sub-loop and `numelem` is set to 10 and `elemsize` to 4 again and initializes the whole array correctly.

With the old code we later do
```
i = *repeat = ad_val_of(AD_NUMELM(AD_DPTR(ict->dtype)));
```
Which initialized both `i` and `*repeat` to `n` (4 in our case). But we have just initialized the whole array with the recursive call and  if we now run again for just 4 elements this is not going to be enough to match the `num_elem=10`, which the algorithm expects. I also do not understand why should we initialize all 10 elements again.

With the new code, we set both `i` and `*repeat` to 10. And later `numelem - i` becomes 0 and we exit the loop. Now that I think about it this solution might be a bit "hacky".... 

This is the code I am running:
```
  IMPLICIT NONE
  INTEGER, PARAMETER :: n=4
  INTEGER, PARAMETER :: m=10 
  INTEGER :: i, j, k
  INTEGER :: array(m) = (/ (((i), j=1,i), i=1,n ) /) !(/1, 2, 2, 3, 3, 3, 4, 4, 4, 4.../)
```

Reply:
@bryanpkc , @kiranchandramohan  would you have a look at [this snippet](https://github.com/flang-compiler/flang/blob/master/tools/flang2/flang2exe/dinit.cpp#L5145-L5157) of code?
It claims to "collapse nested array constructor". I examined the code and it seems to [glue all nested loops together into a single one](https://github.com/flang-compiler/flang/blob/72e87c583df2cb015349adac0702019da7190f6c/tools/flang2/flang2exe/dinit.cpp#L5396). 
This would explain why we get just two levels of nesting in the ict, no matter how many nested loops there are:
```
  INTEGER :: array2(m) = (/ ((i, j=1, i), i=1, n ) /) ! ict has two levels of subict
  INTEGER :: array3(m*o) = (/ (((i, j=1, i), i=1,n ), k=1, o) /)   ! ict has two levels of subict
```
Then the recursion found in [dinit_varref](https://github.com/flang-compiler/flang/blob/72e87c583df2cb015349adac0702019da7190f6c/tools/flang2/flang2exe/dinit.cpp#L581) only triggers once to handle the subict and it always performs all iterations of a "collapsed", or I would rather say "unrolled" or "flattened" nested loop.

So my solution effectively just cuts any further evaluations after `dinit_varref()` was called for `ict->subict`, which makes sense in the view of what has been done to the `ict` earlier.

I would stay with the solution I proposed and just add a comment explaining the situations. 
Do you see any flaw in the thinking above and if not - would the current solution be acceptable?

Reply:
Thanks for the detailed analysis @michalpasztamobica! Is it possible to add a check here, and `assert` or throw ICE if
`ict->subc` isn't "unrolled" as expected?

Reply:
Adding the internal error/warning wasn't trivial as I saw a few runtime functions throwing it during compilation (the `norm2` function, namely). It took me a while to figure out that there's this hidden info under `u1` structure which lets me distinguish an implied-do loop from other cases. The solution I propose is just below. Let me know if this is what you had in mind?
Also - I suggest we leave it as a warning. This code must have been unused or incorrectly used for years and somehow passed all the testing and didn't raise any concerns, so I am afraid we might block someone's otherwise effective work by adding an error here.

Reply:
@bryanpkc , is the Internal warning OK as is?

Reply:
Hi @bryanpkc , I'd be very grateful for your feedback on this, so we can merge this patch and set up the next one, which resolves @pawosm-arm 's issue with casting arrays of different numeric types.

Reply:
Yes, LGTM.